lib/bashlog.sh

```diff
@@ -14,26 +14,27 @@ function _log_exception() {
 export -f _log_exception;
 
 function log() {
-  local date_format="${BASHLOG_DATE_FORMAT:-+%F %T}";
-  local date="$(date "${date_format}")";
-  local date_s="$(date "+%s")";
-
+  local syslog="${BASHLOG_SYSLOG:-0}";
   local file="${BASHLOG_FILE:-0}";
-  local file_path="${BASHLOG_FILE_PATH:-/tmp/$(basename "${0}").log}";
-
   local json="${BASHLOG_JSON:-0}";
-  local json_path="${BASHLOG_JSON_PATH:-/tmp/$(basename "${0}").log.json}";
+  local stdout_extra="${BASHLOG_EXTRA:-0}";
 
-  local syslog="${BASHLOG_SYSLOG:-0}";
-  local tag="${BASHLOG_SYSLOG_TAG:-$(basename "${0}")}";
+  if [ "${file}" -eq 1 ] || [ "${json}" -eq 1 ] || [ "${stdout_extra}" -eq 1 ]; then
+    local date_format="${BASHLOG_DATE_FORMAT:-+%F %T}";
+    local date="$(date "${date_format}")";
+    local date_s="$(date "+%s")";
+  fi
+  local file_path="${BASHLOG_FILE_PATH:-/tmp/${0##*/}.log}";
+  local json_path="${BASHLOG_JSON_PATH:-/tmp/${0##*/}.log.json}";
+
+  local tag="${BASHLOG_SYSLOG_TAG:-${0##*/})}";
   local facility="${BASHLOG_SYSLOG_FACILITY:-local0}";
   local pid="${$}";
 
   local level="${1}";
-  local upper="$(echo "${level}" | awk '{print toupper($0)}')";
+  local upper="${level^^}";
```

Comment:
This doesn't seem to work on Mac. `${level^^}` is introduced in Bash 4, while Mac by default only has Bash 3. Maybe use `tr` instead?

Reply:
Pinkplus, problem is, the performance issues of the whole bashlog.sh were because it's shelling out to external commands...

Reply:
Is it a reasonable compromise to expect people on macos to have a newer bash installed?
I'd expect devs to have a newer bash on macos instead of being stuck in the past.

The scripts take whatever bash is on your path (`#!/usr/bin/env bash`).
Both macports and homebrew can be used to install a newer bash version on MacOS.

Reply:
@pinkplus I tested your suggestion, and it definitely has a perf impact:

```sh
❯ time ./bin/terraform -version
Terraform v1.1.5
on darwin_amd64
./bin/terraform -version  0.16s user 0.17s system 69% cpu 0.474 total
```

With the bash feature:

```sh
❯ time ./bin/terraform -version
Terraform v1.1.5
on darwin_amd64
./bin/terraform -version  0.08s user 0.05s system 83% cpu 0.158 total
```

I think that's a serious perf degradation and the goal of this PR is to improve perf.

Reply:
I understand that the point is to use bash parameter expansion and avoid processes. But this is a problem for Mac users. Mac only bundles bash 3.2. What's more, the default shell was changed to zsh several years ago. I would say it's not normal to expect bash 4, because users generally don't need it.

Replacing one call shouldn't make tfenv much slower. Even with tr it's already much faster than the current implementation.

Reply:
BTW more than 400ms of overhead on terraform calls is insane! I'm using `terraform fmt` in my CI, and that runs on hundreds of directories, that's 100x400ms -> almost 1 extra minute of overhead just so we can say we support people on MacOS who are developers who don't even use macports or homebrew?

Come on, be reasonable about who the target audience of this tool is.

Reply:
I see your point. Just wanted to say that I also use homebrew, but I've never installed bash from homebrew because I don't need it.

I also want to bring up that other similar tools (like pyenv) don't seem to require bash 4. If you decide that tfenv requires bash 4, it will be a surprise.

Reply:
+1, I like the wrapper approach.

Reply:
To save a few more cycles -- not sure how important it is -- can probably conditionally declare one of two wrappers depending on BASH_VERSION.  So it's checked only once, not on each invocation.

Reply:
I'm thinking about solving the wrapper issue by not calling bashlog at all during `terraform` calls unless explicitly required (eg. if you set an env variable to force debug logging, in case you need to debug the shim itself).

I'm currently limited in available cycles, so I'd welcome some collaboration.

Reply:
I can probably help.  I did a bit of experimenting yesterday to optimize the terraform shim codepath as much as possible.

What's the best way to collaborate?  I can send a PR against your PR, or if you want to make me a collaborator on your fork, I can commit directly to the branch.

Reply:
@cspotcode I've made you collaborator on my fork.

Reply:
Thanks, I will add a comment to this pull request to discuss what changes still need to be done.